langchain/src/callbacks/handlers/tracers.ts

```diff
@@ -458,3 +514,158 @@ export class LangChainTracer extends BaseTracer {
+  ): Promise<TracerSession | TracerSessionV2> {
+    const tenantId = this.tenantId ?? (await this.updateTenantId());
+    const configured_endpoint = `${endpoint}&tenant_id=${this.tenantId}`;
+    const response = await fetch(configured_endpoint, {
+      method: "GET",
+      headers: this.headers,
+    });
+    let tracerSession: TracerSessionV2;
+    if (!response.ok) {
+      console.error(
+        `Failed to load session: ${response.status} ${response.statusText}`
+      );
+      tracerSession = {
+        id: uuid.v4(),
+        start_time: Date.now(),
+        tenant_id: tenantId,
+      };
+      this.session = tracerSession;
+      return tracerSession;
+    }
+    const resp = (await response.json()) as TracerSessionV2[];
+    if (resp.length === 0) {
+      tracerSession = {
+        id: uuid.v4(),
+        start_time: Date.now(),
+        tenant_id: tenantId,
+      };
+      this.session = tracerSession;
+      return tracerSession;
+    }
+    [tracerSession] = resp;
+    this.session = tracerSession;
+    return tracerSession;
+  }
+
+  protected async convertRunToV2Run(
+    run: LLMRun | ChainRun | ToolRun
+  ): Promise<RunV2Create> {
+    const session = (this.session ??
+      this.loadDefaultSession()) as TracerSessionV2;
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    let inputs: Record<string, any> = {};
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    let outputs: Record<string, any> | null = null;
+    let child_runs: Array<LLMRun | ChainRun | ToolRun> = [];
+    if (run.type === "llm") {
+      const llmRun = run as LLMRun;
+      inputs = { prompts: llmRun.prompts };
+      outputs = llmRun.response
+        ? {
+            generations: llmRun.response.generations,
+            llmOutput: llmRun.response.llmOutput,
+            [RUN_KEY]: llmRun.response[RUN_KEY],
+          }
+        : {};
+      child_runs = [];
+    } else if (run.type === "chain") {
+      const chainRun = run as ChainRun;
+      inputs = chainRun.inputs;
+      outputs = chainRun.outputs ?? {};
+      child_runs = [
+        ...chainRun.child_llm_runs,
+        ...chainRun.child_chain_runs,
+        ...chainRun.child_tool_runs,
+      ];
+    } else {
+      // run.type === "tool"
+      const toolRun = run as ToolRun;
+      inputs = { input: toolRun.tool_input };
+      outputs = toolRun.output ? { output: toolRun.output } : {};
+      child_runs = [
+        ...toolRun.child_llm_runs,
+        ...toolRun.child_chain_runs,
+        ...toolRun.child_tool_runs,
+      ];
+    }
+
+    return {
+      id: run.uuid,
+      name: run.serialized.name,
+      start_time: run.start_time,
+      end_time: run.end_time,
+      extra: {}, // TODO: VWP
+      error: run.error,
+      execution_order: run.execution_order,
+      serialized: run.serialized,
+      inputs,
+      outputs,
+      session_id: session.id,
+      run_type: run.type,
+      reference_example_id: this.exampleId,
```

Comment:
This isn't right, it would leave the same example as reference example for the root run and all its children runs, it should apply to the root run only

Reply:
That's a great point